### PR TITLE
[WIP] Simplify the image mirroring config

### PR DESCRIPTION
**Pull request by petr-muller**
A probe into simplifying the image mirroring infrastructure under the `core-services`, like this:

## V1(https://github.com/petr-muller/release/tree/mirroring-v1)

```
core-services/quay-mirroring/
core-services/quay-mirroring/knative/jobs.yaml
core-services/quay-mirroring/knative/mapping_...
core-services/quay-mirroring/knative/OWNERS
core-services/quay-mirroring/openshift/jobs.yaml
core-services/quay-mirroring/openshift/mapping_...
core-services/quay-mirroring/openshift/OWNERS
...
```

Advantages:
- People setting up $SUBJ can do so in a single place, not several like before
- No need to touch `Makefile` or `config-updater` config anymore
- CronJobs are covered by the dry-run presubmit and automated application postsubmit
- Top-level approval only needed to set up "areas" (subdirectories), owners self-service afterwards
- New "areas" can be simply created by copying a template directory and few edits

## V2

See https://github.com/openshift/release/pull/4220#issuecomment-507303722

Advantages:
- One-time setup at two places (a presubmit and mappings), afterwards just mappings
- No need to touch `Makefile` or `config-updater` config anymore
- As Prowjobs, they publish their logs and results via Deck
- Top-level approval only needed to set up "areas" (subdirectories), owners self-service afterwards

@stevekuznetsov @smarterclayton WDYT?

If sounds good, the remaining work is:
- [ ] Write the doc describing how this works for the users
- [ ] Adapt the mapping-regenerating targets from existing Makefile
- [ ] Prepare the template directory

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Couple questions:

 - I think we should use Prow periodics instead of CronJobs for these, can we do that switch first before committing this?
 - do we need more than one job? can we just have one large configmap and loop over all possible configs in one job?

**Comment by petr-muller** [comment]
> Couple questions:
> 
>     * I think we should use Prow periodics instead of CronJobs for these, can we do that switch first before committing this?

I'll try that, I don't think there's any reason why these could not be prowjobs. It makes the config more scattered again, but I think it's worth it in this case.

>     * do we need more than one job? can we just have one large configmap and loop over all possible configs in one job?

I was thinking about that myself, and I'd go for one-job-per-area, not one job for everything. These things can fail, and if we had a single one, one failure would be noise for the remaining teams.

**Comment by petr-muller** [comment]
@stevekuznetsov @smarterclayton pushed a new version, PTAL

In this version, the mirroring config consists of:

1. A Prow periodic in `ci-operator/jobs/image-mirroring.yaml` which iterates over mapping files (done with a little inline bash). Failures are reported, but do not prevent the remaining files from processing.
2. A bunch of mapping files under `core-services/quay-mirroring/$AREA` that get loaded into a single CM, so the jobs rely on a name convention to iterate over the correct set of mapping files. It would be possible to use separate config maps and avoid the name convention, but then users would need to one-time setup their config-updater config in `plugins.yaml`, which has the pitfall of needing a separate PR to update that config.

Thoughts?

File: ci-operator/jobs/image-mirroring.yaml [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+              set -o errexit
+              failures=0
+              for mapping in /etc/imagemirror/mapping_knative*; do
+                echo "Running: oc image mirror -f=$mapping --skip-multiple-scopes"
+                if ! oc image mirror -f="$mapping" --skip-multiple-scopes; then
+                  echo "ERROR: Failed to mirror images from $mapping"
+                  failures=$((failures+1))
+                fi
+              done
+              exit $failures
```
Comment (stevekuznetsov) [inline ci-operator/jobs/image-mirroring.yaml]: clever

File: ci-operator/jobs/image-mirroring.yaml [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        - name: config
+          configMap:
+            name: image-mirror-mappings
+  - agent: kubernetes
+    labels:
+      ci.openshift.io/role: image-mirroring
+      ci.openshift.io/area: tekton
+    cron: "@midnight"
+    decorate: true
+    name: periodic-image-mirroring-tekton
```
Comment (stevekuznetsov) [inline ci-operator/jobs/image-mirroring.yaml]: Isn't tekton part of knative?

Comment (petr-muller) [reply ci-operator/jobs/image-mirroring.yaml]: I'm not sure about their relation, but they set up mirroring config separately (I have just followed the existing jobs)

Comment (stevekuznetsov) [reply ci-operator/jobs/image-mirroring.yaml]: weird

**Comment by petr-muller** [comment]
> How many places do we have infrastructure-periodics.yaml hard-coded as the only job file not in the normal dir hierarchy? Do we need to update any tools?

@stevekuznetsov I'm not sure, but I'll investigate. I don't think there are many places like this.

**Comment by petr-muller** [comment]
So if the PoC seems fine, let's transition to it in a slightly less risky way:

1. Merge the config-updater config change
2. Merge the new jobs and mapping files, run the mirroring in dry-mode alongside the current CronJobs
3. See the Prowjobs doing what they should
4. Flip the Prowjobs to non-dry and remove the current CronJobs
5. Communicate the change on aos-devel

/close

**Pull request closed without merging**